app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/StaticPlayingQueue.java

```diff
@@ -0,0 +1,458 @@
+    public static final int REPEAT_MODE_ALL = 1;
+    public static final int REPEAT_MODE_THIS = 2;
+
+    private int repeatMode;
+
+    public static final int INVALID_POSITION = IndexedSong.INVALID_INDEX;
+    public static final int SHUFFLE_MODE_NONE = 0;
+    public static final int SHUFFLE_MODE_SHUFFLE = 1;
+    private int shuffleMode;
+
+    private int currentPosition;
+
+    private int nextPosition;
+
+    private boolean songsIsStale;                     // used to know when songs and queue.song are not equal so that getPlayingQueueSongOnly() will not always recreate songs every time
+    private final ArrayList<Song> songs;              // necessary as all implementation use MusicPlayerRemove.getPlayingQueue and want a pointer that doesn't change to a list of song
+    private ArrayList<IndexedSong> queue;             // list of element currently saved (way better than songs to ensure only the correct occurrence of a song is modified)
+    private final ArrayList<IndexedSong> originalQueue; // copy of the queue used to allow revert of history last operation
+
+
+    public StaticPlayingQueue() {
+        songs = new ArrayList<>();
+        queue = new ArrayList<>();
+        originalQueue = new ArrayList<>();
+        shuffleMode = SHUFFLE_MODE_NONE;
+        currentPosition = INVALID_POSITION;
+        songsIsStale = false;
+    }
+
+    public StaticPlayingQueue(ArrayList<IndexedSong> restoreQueue, ArrayList<IndexedSong> restoreOriginalQueue, int restoredPosition, int shuffleMode) {
+        this.queue = new ArrayList<>(restoreQueue);
+        this.originalQueue = new ArrayList<>(restoreOriginalQueue);
+        this.shuffleMode = shuffleMode;
+
+        currentPosition = restoredPosition;
+
+        songs = new ArrayList<>();
+        songsIsStale = true;
+        resetSongs();
+    }
+
+    public void restoreMode(int shuffleMode, int repeatMode) {
+        this.shuffleMode = shuffleMode;
+        this.repeatMode = repeatMode;
+    }
+
+    /* -------------------- queue modification (add, remove, move, ...) -------------------- */
+
+    // add song at the end of both list
+    public void add(Song song) {
+        queue.add(new IndexedSong((Song)song, queue.size()));
+        originalQueue.add(new IndexedSong((Song)song, originalQueue.size()));
+
+        songsIsStale = true;
+    }
+
+    // add list of song at the end of both list
+    public void addAll(@NonNull List<Song> songs) {
+        for (Song song : songs) {
+            add(song);
+        }
+    }
+
+    // add song after and including position, numbering need to be redone for every song after this position (+1)
+    public void addAfter(int position, Song song) {
+        if (position >= queue.size()) {
+            position = queue.size() - 1;
+        }
+        int previousPosition = queue.get(position).index + 1;
+        position = position + 1;
+
+        originalQueue.add(previousPosition, new IndexedSong(song, previousPosition));
```

Comment:
`new IndexedSong(song, previousPosition)` can be extracted to a variable as it is reused 3 times